core/src/test/java/org/apache/calcite/test/RelBuilderTest.java

```diff
@@ -5585,6 +5590,284 @@ private static RelNode buildCorrelateWithJoin(JoinRelType type, RelBuilder build
+    RelOptCost filteredCost = mq.getCumulativeCost(filtered);
+    RelOptCost aggregatedCost = mq.getCumulativeCost(aggregated);
+
+    assertThat(scanCost, notNullValue());
+    assertThat(filteredCost, notNullValue());
+    assertThat(aggregatedCost, notNullValue());
+
+    // Get Combine's costs
+    RelOptCost combineSelfCost =
+        combined.computeSelfCost(combined.getCluster().getPlanner(), mq);
+    RelOptCost combineCumulativeCost = mq.getCumulativeCost(combined);
+
+    assertThat(combineSelfCost, notNullValue());
+    assertThat(combineCumulativeCost, notNullValue());
+
+    // Verify cumulative cost is sum of all input costs plus self cost
+    RelOptCost expectedTotal = combineSelfCost
+        .plus(scanCost)
+        .plus(filteredCost)
+        .plus(aggregatedCost);
+
+    // The cumulative cost should equal the sum
+    assertThat(combineCumulativeCost.getRows(),
+        is(expectedTotal.getRows()));
+    assertThat(combineCumulativeCost.getCpu(),
+        is(expectedTotal.getCpu()));
+    assertThat(combineCumulativeCost.getIo(),
+        is(expectedTotal.getIo()));
+
+    // Verify relationships between costs
+    // Filtered should cost more than simple scan
+    assertThat(filteredCost.isLt(scanCost), is(false));
+
+    // Aggregated should cost more than simple scan
+    assertThat(aggregatedCost.isLt(scanCost), is(false));
+
+    // Combined cumulative cost should be greater than any individual cost
+    assertThat(combineCumulativeCost.isLt(scanCost), is(false));
+    assertThat(combineCumulativeCost.isLt(filteredCost), is(false));
+    assertThat(combineCumulativeCost.isLt(aggregatedCost), is(false));
+  }
+
+  @Test void testCombineCumulativeCostWithSharedInputs() {
+    // Test cumulative cost when Combine has shared subexpressions
+    final RelBuilder builder = RelBuilder.create(config().build());
+
+    // Create a base scan that will be shared
+    RelNode baseScan = builder.scan("EMP").build();
+
+    // Create two queries that use the same base scan
+    RelNode query1 = builder.push(baseScan)
+        .filter(builder.equals(builder.field("DEPTNO"), builder.literal(10)))
+        .project(builder.field("ENAME"))
+        .build();
+
+    RelNode query2 = builder.push(baseScan)
+        .filter(builder.equals(builder.field("DEPTNO"), builder.literal(20)))
+        .aggregate(builder.groupKey(), builder.count())
+        .build();
+
+    // Combine the queries
+    RelNode combined = builder.combine(query1, query2).build();
+
+    RelMetadataQuery mq = combined.getCluster().getMetadataQuery();
+
+    // Get costs
+    RelOptCost query1Cost = mq.getCumulativeCost(query1);
+    RelOptCost query2Cost = mq.getCumulativeCost(query2);
+    RelOptCost combinedCost = mq.getCumulativeCost(combined);
+
+    assertThat(query1Cost, notNullValue());
+    assertThat(query2Cost, notNullValue());
+    assertThat(combinedCost, notNullValue());
+
+    // Even with shared base scan, Combine's cumulative cost
```

Comment:
Q: is the cost of the shared subtree duplicated here? Should it be?

I think the answer to the aforementioned questions should emerge clearly from the test

Reply:
Yes, I think so. We’d want the cost of a `Combine` plan that doesn’t identify the obvious shared expression to be higher than that of a plan that pulls the shared query out and uses `Spool` (or another mechanism) to lower the overall cost.

Things get trickier when rules have already transformed sub-expressions that could have been shared. For example, a filter pushdown in one branch might lower its individual cost but prevent it from being recognized as a common expression. In that scenario, we’d still expect the plan that uses a single scan -> filter and `Spool` to have the lower overall cost.